### DISCUSSION TITLE
Print table, chart or map

### DISCUSSION THREAD

**Discussion by WohnungNDS** [open-sdg/open-sdg#1436] [Ideas]
**Is your feature request related to a problem? Please describe.**
My clients like printing very well. They want to Print the table, the chart or even the map.
@MoBosse may be your clients like printing everything as well :-)

**Describe the solution that you need**
next to the "download Button I'd like to have a Button for printing the "main content"

I found this js on the web and added it to my [custom.js](https://github.com/WohnungNDS/IM_Site/blob/develop/assets/js/custom.js) file
````
function printDiv() 
{

  var divToPrint=document.getElementById('main-content');

  var newWin=window.open('','Print-Window');

  newWin.document.open();

  newWin.document.write('<html><body onload="window.print()">'+divToPrint.innerHTML+'</body></html>');

  newWin.document.close();

  setTimeout(function(){newWin.close();},10);

}
````
Then I put this line 

`<input type='button' id='btnSave' value='Print' class="btn btn-primary btn-download" onclick='printDiv();'>`

in my [previous_reports.md](https://github.com/WohnungNDS/IM_Site/blob/develop/_pages/previous_reports.md)

and it works great as you can see [here](https://wohnungnds.github.io/IM_Site/previous_reports/)

clicking on the print-button the printing dialoge opens.....

![image](https://user-images.githubusercontent.com/67795376/140277996-3d85b752-6062-47ea-9daa-6b10f8211d89.png)

Two Questions:

1) Where do I put this line to print out the table, chart or map?
2) I am not familiar to .js. It looks like a new window (with the content tp print) will be opened. After printing the window will be closed. It will be more smart to print the div directly. How do I have to change the js file to do this?

**Comment by MoBosse** [comment]
@WohnungNDS in fact I am just thinking about a similar thing to implement on our [national indicators](https://sustainabledevelopment-deutschland.github.io/en/) platform. I would like to create a button that allows the user to download a pdf with all the indicator information, the Chart and the text passages from the metadata tabs. 
I have used html2pdf() for this, but am having some problems getting a decent layout. Anyway, to get the chart into the PDF, I used `document.getElementById('chart-canvas')` instead of all the content as you did above with `main-content`.

**Comment by WohnungNDS** [comment]
@MoBosse sounds great.... some kind of fact sheet for an indicator?

But the first question is where do I have to change my code that the button appears next to the other buttons into my indicator site?
To my mind it has to be possible to fill 

`var divToPrint=document.getElementById('**main-content**');`

with the current div that is shown on the indicator page...

for the table It might be this: `<div id="selectionsTable">`
for the chart It might be this: `<div id="selectionsChart">`
for the map It might be this: `<div id="mapview">`

But i am not familar with the html strucure of the indicator page.....

perhaps @LucyGwilliamAdmin or @brockfanning givew a hint ;-)
.

**Comment by MoBosse** [comment]
My first thought was that you would have to overwrite the [indicatorView ](https://github.com/open-sdg/open-sdg/blob/1.7.0-dev/_includes/assets/js/indicatorView.js) file for this because here is the `createDownloadButton` function that creates the green button underneath the chart aswell, but I am not sure about that.

**Comment by WohnungNDS** [comment]
@MoBosse may be..... but as I said before i am not familiar with js ...

perhaps we get a hint from the other users....

**Comment by christoph-LSN** [comment]
@MoBosse @brockfanning 

We have now implemented this on our platform for **tables** and **maps**.

[My Site](https://christoph-lsn.github.io/MT_Site/1-1-1/)


<img width="1571" height="397" alt="image" src="https://github.com/user-attachments/assets/06eedd4e-e926-4eaf-9b67-8db19a399e27" />


<img width="1447" height="811" alt="image" src="https://github.com/user-attachments/assets/3b26d84c-5c38-4a43-a475-6b6b88e07bd2" />


### Changed files
- [map-print.js](https://github.com/christoph-LSN/MT_Site/blob/develop/_includes/assets/js/map-print.js)
- [table-export.js](https://github.com/christoph-LSN/MT_Site/blob/develop/_includes/assets/js/table-export.js)
- [table-print.js](https://github.com/christoph-LSN/MT_Site/blob/develop/_includes/assets/js/table-print.js)
- [custom.scss](https://github.com/christoph-LSN/MT_Site/blob/develop/_sass/custom.scss)
-  [scripts-custom.html](https://github.com/christoph-LSN/MT_Site/blob/develop/_includes/scripts-custom.html)

### What we added

#### 1. Table printing
We added a dedicated table print button that stays available in the table area and continues to work after:
- changing area selections
- changing visible columns
- sorting the table

The print view was cleaned up so that it shows:
- one clear main heading
- the table
- a source line

We also added some print-specific spacing and layout styling.

#### 2. Table XML export
In addition to print, we added a second button that exports the **currently visible table state** as XML.

That means the export respects the table exactly as the user has configured it, including:
- current area selection
- current sorting
- current filtering / search
- currently visible columns

We switched this to an **Excel-friendly SpreadsheetML XML export**, so that opening the file in Excel gives a worksheet-style result instead of a raw XML tree.

#### 3. Map printing
We added a dedicated print button for the map as well.

The printed map now includes:
- a clean title block
- the currently selected reference year
- a source line
- the legend in the print layout

We also handled the legend and tooltip behavior more carefully:
- when nothing is selected, the min/max legend remains visible
- when regions are selected, the value tooltips can remain visible in the print output so the printed circles can still be matched to their values

### Current scope
At the moment this covers:
- **table print**
- **table XML export**
- **map print**

We have **not** implemented chart printing yet.

If helpful, I can also share the relevant snippets or open a follow-up example for the DataTables / Leaflet integration.

I am not familar with JS Copilot helped me to create and implemt this functions - so I would be glad if you tae a look...

Christoph

**Comment by christoph-LSN** [comment]
Hi @LucyGwilliamAdmin may cou can take a look at this?

**Comment by sarah-beasley** [comment]
Hi @christoph-LSN , thank you for putting this together, it looks great! I am also not a JS expert, but hopefully @brockfanning can comment on any of the specifics.

**Comment by MoBosse** [comment]
Hi Christoph, it looks great!! Unfortunately, I can’t help with the programming either, but I could offer to have me (or a colleague) test the functions...
For example, I’ve noticed that the units/disaggregation categories aren’t included on the map and table printouts, so sometimes you don’t know exactly what the printout represents. [Here](https://christoph-lsn.github.io/MT_Site/2-1-1/ ), for example, you can’t tell from the map whether a gender has been selected.
Let me know if this kind of feedback would be helpful to you.

**Comment by christoph-LSN** [comment]
@MoBosse Thank you for looking into it. Choosen units or series are shown in printout the subtitle.
Anzahl:
<img width="450" height="252" alt="{88E9A653-C3D5-466A-8E83-0286E479052A}" src="https://github.com/user-attachments/assets/6c610446-80b7-4885-8a71-34241e6abc61" />
Prozent:
<img width="1020" height="247" alt="{4DDD20C5-8061-4EE5-B553-062D5F3122DD}" src="https://github.com/user-attachments/assets/27003852-1a8b-4b56-add1-fa8bf7a10d08" />

For the map this may an option too.... (I' look into it an get back to you)

An yes I'm looking for testing people as well as for someone who looks into my code

**Comment by christoph-LSN** [comment]
@MoBosse Now the chossen disaggregations in the map view will be printed also...

<img width="1210" height="337" alt="{36031F60-D621-4949-8631-F52675E8DCF8}" src="https://github.com/user-attachments/assets/00da05ad-421e-4326-a115-1b3135e73565" />
